src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Formatting/FormattingLanguageServerClient.cs

```diff
@@ -105,6 +115,48 @@ private RazorDocumentRangeFormattingResponse Format(RazorDocumentRangeFormatting
+
+                var textBufferFactoryService = new FakeTextBufferFactoryService(contentTypeService);
+                var textBufferListeners = Array.Empty<Lazy<IWebTextBufferListener, IOrderedComponentContentTypes>>();
+                var bufferManager = Activator.CreateInstance(bufferManagerType, new object[] { contentTypeService, textBufferFactoryService, textBufferListeners });
+                var threadSwitcher = (IThreadSwitcher)Activator.CreateInstance(
+                    Assembly.Load("Microsoft.WebTools.Shared, Version=16.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a").GetType("Microsoft.WebTools.Shared.Threading.JoinableTaskFactoryThreadSwitcher", throwOnError: true),
+                    new object[] { new JoinableTaskContext().Factory });
+                var applyFormatEditsHandler = Activator.CreateInstance(editHandlerType, new object[] { bufferManager, threadSwitcher, textBufferFactoryService });
+
+                // Make sure the buffer manager knows about the source document
+                var documentUri = DocumentUri.From($"file:///{@params.HostDocumentFilePath}");
+                var contentTypeName = HtmlContentTypeDefinition.HtmlContentType;
+                var initialContent = generatedHtml;
+                var snapshotVersionFromLSP = 0;
+                Assert.IsAssignableFrom<ITextSnapshot>(bufferManager.GetType().GetMethod("CreateBuffer").Invoke(bufferManager, new object[] { documentUri, contentTypeName, initialContent, snapshotVersionFromLSP }));
+
+                var requestType = editHandlerAssembly.GetType("Microsoft.WebTools.Languages.LanguageServer.Server.ContainedLanguage.ApplyFormatEditsParamForOmniSharp", throwOnError: true);
+                var serializedValue = $@"{{
+    ""Options"": {{
+        ""UseSpaces"": {(@params.Options.InsertSpaces ? "true" : "false")},
+        ""TabSize"": {@params.Options.TabSize},
+        ""IndentSize"": {@params.Options.TabSize}
+    }},
+    ""Uri"": ""file:///{@params.HostDocumentFilePath}"",
+    ""GeneratedChanges"": [
+    ]
+}}
+";
+                var request = JsonConvert.DeserializeObject(serializedValue, requestType);
+                var resultTask = (Task)applyFormatEditsHandler.GetType().GetRuntimeMethod("Handle", new Type[] { requestType, typeof(CancellationToken) }).Invoke(applyFormatEditsHandler, new object[] { request, CancellationToken.None });
+                var result = resultTask.GetType().GetProperty(nameof(Task<int>.Result)).GetValue(resultTask);
+                var rawTextChanges = result.GetType().GetProperty("TextChanges").GetValue(result);
+                var serializedTextChanges = JsonConvert.SerializeObject(rawTextChanges, Newtonsoft.Json.Formatting.Indented);
+                var textChanges = JsonConvert.DeserializeObject<HtmlFormatterTextEdit[]>(serializedTextChanges);
+                response.Edits = textChanges.Select(change => change.AsTextEdit(SourceText.From(generatedHtml))).ToArray();
```

Comment:
📝 If we can't get this working in CI in a timely manner, the other option is to add a _record_ functionality to match the current _replay_ functionality. In record mode, a developer must provide the necessary dependencies, and instead of reading/using the baseline inputs and outputs, each test creates or updates the baselines using real data from the formatter.